### DISCUSSION TITLE
nextjs-ts-user-management, how to handle redirect to login page using @supabase/auth-helpers-nextjs  ?

### DISCUSSION THREAD

**Discussion by Jared-Dahlke** [supabase/supabase#11085] [General]
First of all wow, i love Supabase thank you for building such an awesome platform. I'm enjoying learning it as I build a project.

I'm leveraging `https://github.com/supabase/supabase/tree/master/examples/user-management/nextjs-ts-user-management`
to build basically a TODO/hello world app and it's amazing. But you will notice that the way they handle loginpage/protected pages toggle in [nextjs-ts-user-management](https://github.com/supabase/supabase/blob/master/examples/user-management/nextjs-ts-user-management/pages/index.tsx#L13) is via `session ? <Account/>: <Login/>` which is not compatible with using Next's file directory routing system. 

I have successfully implemented a `middleware.ts` to redirect the user to login page if they're not authenticated per the instructions here which works great! https://supabase.com/docs/guides/auth/auth-helpers/nextjs#auth-with-nextjs-middleware
but it only works when user tries to navigate to a new page.

The middleware.ts does not redirect the user for the below situations:
1. user clicks a `signOut` button
2. user revisits the page after he is no longer authenticated

Is there some best practice way of addressing those 2 situations since it doesnt seem possible to do using middleware?

I've tried making  "ProtectedRoutes" and stuff like that but it has 2 problems (1.no way to see if session is loading so login page flashes on page load, 2. it doesnt actually change the url, they are still on /profile but seeing the login page for example.)
![image](https://user-images.githubusercontent.com/32936978/208312011-4cffd776-cbba-49b8-b7fc-979bc7678fbe.png)

any ideas or help would be much appreciated.

**Comment by xvvvyz** [comment]
I think this article explains a possible solution:

https://supabase.com/docs/guides/auth/auth-helpers/nextjs-server-components#supabase-listener
